website/static/js/filters.js

```diff
@@ -0,0 +1,149 @@
+                                }
+                inputs:     -a dictionary of the search fields
+                            -key: the ID for the search field
+                            -value: the selector that this field should search against
+         */
+        var settings = $.extend({
+            items: ['.items']
+        }, options);
+
+        var itemsSelector = settings.items.join();
+
+        var filterConstraints = {};
+        var searchConstraints = {};
+
+        var search = function(el) {
+            for (var key in searchConstraints) {
+                if (searchConstraints.hasOwnProperty(key)) {
+                    var content = $(el.querySelector(key)).text().toLowerCase();
+                    var exists = content.indexOf(searchConstraints[key]);
+                    if (exists === -1) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        };
+
+        var filter = function(el) {
+            for (var key in filterConstraints) {
+                if (filterConstraints.hasOwnProperty(key)) {
+                    var selector = settings.groups[key].filter;
+                    var type = settings.groups[key].type;
+                    var match = filterConstraints[key];
+                    if (type === 'text') {
+                        var content = $(el.querySelector(selector)).text();
+                        if (match.indexOf(content) === -1) {
+                            return false;
+                        }
+                    }
+                    else if (type === 'checkbox') {
+                        if (match.indexOf(el.querySelector(selector).checked) === -1) {
+                            return false;
+                        }
+                    }
+                }
+            }
+            return true;
+        };
+
+        var toggle = function() {
+            var activeItems = $(itemsSelector);
+            $(itemsSelector).each(function() {
+                var self = this;
+                if (!search(self) || !filter(self)) {
+                    activeItems.splice(activeItems.index(self), 1);
+                    $(self).fadeOut();
+                }
+            });
+            $(activeItems).fadeIn();
+            if (settings.callback !== undefined) {
+                var empty = [];
+                var filtered = [];
+                for (var i = 0, selector; selector = settings.items[i]; i ++) {
+                    var active = $(activeItems).filter(selector);
+                    var items = $(selector);
+                    if (active.length < items.length) {
+                        filtered.push(selector);
+                    }
+                    if (active.length === 0) {
+                        empty.push(selector);
+                    }
+                }
+                settings.callback(filtered, empty);
+            }
+        };
+
+        var clickFunction = function() {
```

Comment:
Oh one more thing, can you add docstrings to these functions? I feel like there's a lot happening here so it would be useful to have a quick way of knowing what each function is doing.